### PR TITLE
Improve sorting in dataframe editor

### LINKED ISSUE #3010: DataFrame editor should do a stable sort
I am using Spyder 2.3.5.2 from WinPython-64bit-3.4.3.5.

I noticed that clicking column headers in the DataFrame editor does not sort the table [in a stable way](https://en.wikipedia.org/wiki/Sorting_algorithm#Stability). This especially means that one cannot reach a hierarchical sorting.

### LINKED ISSUE #3020: DataFrame editor should first sort up, then down
## Description of your problem

When clicking a column header in the DataFrame editor, the column is sorted from high to low. A second click sorts it from low to high. Since sorting from low to high is needed much more often, the effects of the first and the second click should be reversed.

**What steps will reproduce the problem?**

1. Open a Pandas DataFrame in the editor.
2. Click on a column header.

**What is the expected output?**
The DataFrame column is sorted s.t. the lowest value is on top.

**What do you see instead?**
The DataFrame column is sorted s.t. the lowest value is on bottom.

**Please provide any additional information below**
N/A

## Versions and main components

* Spyder Version: 2.3.5.2 from WinPython-64bit-3.4.3.5.
* Python Version: 3.4.3
* Operating system: Windows 7 64-bit

## Dependencies

Please go to the menu entry `Help > Optional Dependencies` (or
`Help > Dependencies`), press the button `Copy to clipboard`
and paste the contents below:

IPython >=1.0   :  3.2.1 (OK)
jedi >=0.8.1    :  0.9.0 (OK)
matplotlib >=1.0:  1.4.3 (OK)
pandas >=0.13.1 :  0.16.2 (OK)
pep8 >=0.6      :  1.6.2 (OK)
pyflakes >=0.6.0:  0.9.2 (OK)
pygments >=1.6  :  2.0.2 (OK)
pylint >=0.25   :  1.4.3 (OK)
rope >=0.9.2    :  0.9.4-1 (OK)
sphinx >=0.6.6  :  1.3.1 (OK)
sympy >=0.7.3   :  0.7.6 (OK)
zmq >=2.1.11    :  14.7.0 (OK)

### PR DESCRIPTION
**Pull request by jitseniesen**
Dataframe editor: Make sorting stable; handle order correctly.

Fixes #3010 
Fixes #3020

### CODE REVIEW

File: spyderlib/widgets/variableexplorer/tests/test_dataframeeditor.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    assert dfm.data(dfm.createIndex(0, 2)) == 'c'
+    assert dfm.data(dfm.createIndex(1, 0)) == '1'
+    assert dfm.data(dfm.createIndex(1, 1)) == '3'
+    assert dfm.data(dfm.createIndex(1, 2)) == 'a'
+    
+def test_dataframemodel_sort():
+    df = DataFrame({'colA': [1, 3], 'colB': ['c', 'a']})
+    dfm = DataFrameModel(df)
+    dfm.sort(2)
+    print(dfm.df)
```
Comment (Nodd) [inline spyderlib/widgets/variableexplorer/tests/test_dataframeeditor.py]: :no_entry_sign:

**Comment by Nodd** [comment]
The code looks good to me. And it even has tests ! @jitseniesen you passed the 10% coverage :smile:

File: spyderlib/widgets/variableexplorer/tests/test_dataframeeditor.py (hunk 1 above)
Comment (ccordoba12) [inline spyderlib/widgets/variableexplorer/tests/test_dataframeeditor.py]: Yep, please remove it :-)

**Comment by jitseniesen** [comment]
> The code looks good to me. And it even has tests ! @jitseniesen you passed the 10% coverage :smile:

I like tests, because I make many mistakes!

File: spyderlib/widgets/variableexplorer/tests/test_dataframeeditor.py (hunk 1 above)
Comment (jitseniesen) [inline spyderlib/widgets/variableexplorer/tests/test_dataframeeditor.py]: Oops, I hadn't spotted that I left the print in. It's gone now.

**Comment by ccordoba12** [comment]
What a nice addition @jitseniesen!! Thanks a lot for helping us with it and for increasing our testing coverage too!

----

*Note*: I saw you rebased your PR to fix failures in Travis. Usually those failures are caused by random segfaults or tracebacks, so there's no need to worry about that. I usually fix it by launching new builds before merging :-)

However, if you want to trigger a new build in Travis and AppVeyor yourself, you can easily do it by closing and opening the PR (instead of rebasing :-).

**Comment by goanpeca** [comment]
Awesome :-)

Thanks a lot @jitseniesen

**Comment by jitseniesen** [comment]
> However, if you want to trigger a new build in Travis and AppVeyor yourself, you can easily do it by closing and opening the PR (instead of rebasing :-).

Thanks, that's good to know. It is a bit annoying that the builds fail that often.

**Comment by goanpeca** [comment]
>Thanks, that's good to know. It is a bit annoying that the builds fail that often.

Yes it is!!!, but sometimes the segfaults in tests... "seem" to be random :-|

**Pull request merged**